design-documents/asynchronous-operations/consumer-processes-improvements.md

```diff
@@ -0,0 +1,87 @@
+That should save some precious server resources.
+
+### Making these options configurable and have some defaults
+
+These options should be configurable per consumer type.
+Some sensible defaults could be set in the [`queue_consumer.xml`](https://devdocs.magento.com/guides/v2.3/extension-dev-guide/message-queues/config-mq.html#queueconsumerxml) file for some of these options.  
+Next to that, developers or shopowners should be able to override these values per consumer type. At least being able to override them in the `app/etc/env.php` file would be nice, but a backend interface for making these things configurable would also be very nice (but can maybe be done in a later phase?).
+
+### Credits
+
+Some of these ideas where taken from the [`symfony/messenger` component](https://github.com/symfony/messenger/blob/3d65f22f9a56f6475c19999fdbc3a897cefc8900/Command/ConsumeMessagesCommand.php#L77-L80).
+
+## Problem 2: deployment problems
+
+### The problem
+
+When deploying a new version of Magento/third party/custom code to a server, we can potentially run against a problem where consumers are still using old code loaded in memory. This is not only the code the consumer is running itself, but might also be Magento core code itself for the consumer processes themselves (which might change when upgrading Magento to a newer version).
+
+Also when using a capistrano-style deployment where you have a symlinked a `current` directory pointing to a certain release directory, after a new deploy the running consumers will still supposedly run from the old release directory referencing a `.pid` file containing its process id in a directory which is in the old release directory. The cronjob `consumers_runner` will go searching for that file in the new release directory and won't find it. Causing a second consumer to start up, even though the old one is still running. (This is probably already fixed by this unreleased commit: [MC-18477](https://github.com/magento/magento2/commit/1d9e07b218c7c8ad1f05706828cb2dd47d2d2d58))
+
+### The suggested solution
+
+The suggestion here would be to give deployment scripts the option to signal the running consumer proccesses to handle their current messages and then kill themselves as soon as possible.  
+We already have the poison pill functionality which could be used here, but the problem with that one is that it will only be checked when a new message appears in the queue. If no messages appear in the queue during a deploy, and we only update the poison pill version, the consumer will still keep running because it doesn't check for the poison pill version until a new message appears.  
+So I would suggest some kind of dummy message to be created in each queue which only purpose is for the consumers to check the poison pill version and that message then should get removed again from the queue and nothing should be done with it.
```

Comment:
That's true on the one hand, but on the other, there maybe cases where you explicitly want to stop consumer processes at a specific time, for example during a deploy to put a new version of the code online.
I know some people stop the crontab during deploys, so when you kill consumers using this new command, and crons are disabled, no new ones will spawn until the crons are enabled again after the deploy was finished.
It "feels" more clean to do a deploy in this way.

But personally I'm fine with leaving this out if it's not wanted.

Reply:
Leaving this in here for now, it doesn't sound like there is a consensus reached yet.
Let me know if I misunderstood this.

Reply:
let me confirm and get to some consensus with @nuzil ; the overall approach seems to work fine, just want to make sure, we are all on same page.

Reply:
@hostep please go-ahead and update, we are in agreement. Since it's configurable, and the integrator can decide if they want to enable this functionality for 'consumers_runner' process. Optionally when max-idle-time is not set at all then it will revert to default behavior; which may give more flexibility to integrator.

Reply:
Hi @tariqjawed83 

Sorry, I'm a little bit confused, are we still talking about keeping or removing this "sending new dummy message to queue" feature?
And what should I update exactly?

Tnx!

Reply:
@hostep as discussed above, this dummy message to queue feature may not be required with the other options in-place, as below
 
* only-spawn-when-message-available
* max-idle-time 
* consumers_wait_for_messages
 
consumers_wait_for_messages when set to false, then consumer process will stop the moment it doesn't find any new messages. Which means we don't need any special signaling in that case. Which also means, we may be able to remove the max-idle-time flag from requirements? and just these two options will do the job and solve both long running processes concerns and stopping immediately without waiting for new messages

* only-spawn-when-message-available
* consumers_wait_for_messages

Since we want to make sure, the process should gracefully shutdown, the existing poison pill feature, along with the above two option should work. Let me know if it's not clear yet?

Reply:
I'll just remove the entire Problem 2 section, it looks like there is no interest in this. Even though I still believe certain persons would find it useful.

Let's just keep the proposal for `max-idle-time` and `only-spawn-when-message-available`.

I do have my opinion around that new `consumers_wait_for_messages` flag, it will unnecessarily spawn a `bin/magento` command which tends to take a lot of memory just to bootstrap. I would therefore like to avoid this, so I'm in favor of the `only-spawn-when-message-available` to prevent consumers to spawn at all.
Otherwise, every single minute a new consumer process will spawn, taking a bunch of memory and then destroy itself shortly after if no message was in the queue, this feels wasteful.

Also, suppose a shop owner is using one of these async features (let's say the 'Mass editing of products' thing), it can be assumed that he'll try to use it a couple of times in a short time period. Therefore I would like to have the consumer process to run for let's say 30 minutes, as it will then pickup messages immediately after they enter the queue and don't have to wait on the cron to spawn a new process. In this case the `max-idle-time` is more suited over the `consumers_wait_for_messages` flag in my opinion.

Reply:
Sorry, I didn't realize we still need to solve problem 2 for people not wanting to use the `max-idle-time` flag. So I've added it again and removed the dummy message part.

Is this ok now?

Reply:
@hostep I would agree with you on this, 'max-idle-time' would be more flexible approach to dispose the currently running queue process; If its sets to '0' then it could potentially act like no waiting; so it gives flexibility to the site administrator.

Reply:
Indeed.

Also, be aware that the proposal is for these configuration options to exist per consumer type. The new `consumers_wait_for_messages` configuration option on the other hand is now implemented to act globally for *all* consumers, which is also not flexible enough in my opinion.

So it looks like most of it is now clear and we are on the same page, what are the next steps to move this proposal forward?

Thanks!

Reply:
@hostep I can help this move forward, although wanted to check how we are planning to implement this feature as part of ConsumersRunner job, how much refactoring this may require? wanted to do  quick feasibility check, since ConsumersRunner is not dealing with messages directly as of now. Do you have some strategy in mind?

Reply:
Hi @tariqjawed83 

Thanks for bringing this up during the weekly architectural meeting, you explained it very well! 🙂 

As for the implementation details, I haven't looked into that yet since I'm not super familiar with the code (yet). If you do want some help looking into this, let me know (I'm also available on Slack most of the time if that makes it easier to communicate).